### ISSUE TITLE
Is the watchface functional ?

### ISSUE THREAD

**Issue by Gby56** [rootasjey/feels-fitbit#1]
After taking some clues from your API implementation, I was wondering if this was even functional, but it seems like it isn't.

I was trying to implement DarkSky integration with https://github.com/robro/FitbitArcs , but can't really figure out how to fetch data.
Should it be fetched by the Versa/Ionic, or by the companion ? Should it be fetched regularly and cached or fetched only when asked to display ? (Caching the previous while waiting for fetch then update the UI)

If you can help around, I would be interested to know what's the best way to implement this, I don't understand fully how fetch works, if the Versa calls it, I supposed the phone would fetch then send through websocket the result, but I'm not sure about it.

Thanks :)

**Comment by rootasjey** [comment]
You're right.
I started this project but haven't get it work yet, sorry.

You should see my other project [metrix-fitbit](
https://github.com/rootasjey/metrix-fitbit) where I fetch and show weather's data from DarkSky.
You can deploy the clock face and see it working.

To fetch weather data, the straightforward way is to use [fitbit-weather](https://github.com/gregoiresage/fitbit-weather) and to follow its documentation (readme.md). It already setup the necessary endpoints but you could do it manually.

>Should it be fetched by the Versa/Ionic, or by the companion ?

Versa/Ionic cannot fetch directly data.
To fetch data through Web APIs, you need a [companion](https://dev.fitbit.com/build/guides/companion/). The companion is run on the mobile device. When the  companion got the data, it has to pass it to the watch through the [messaging API](https://dev.fitbit.com/build/guides/communications/messaging/).

Ionic/Versa   | Companion (mobile) |
-------------|---------------------|
Ask for fetch through [Messaging API](https://dev.fitbit.com/build/guides/communications/messaging/) | [Fetch](https://dev.fitbit.com/build/reference/companion-api/fetch/) data and respond to Ionic/Versa through Messaging API
On device [example](https://github.com/rootasjey/metrix-fitbit/blob/master/app/activities/weather.js) | On mobile [example](https://github.com/rootasjey/metrix-fitbit/blob/master/companion/index.js) with [fitbit-weather](https://github.com/rootasjey/metrix-fitbit/blob/83765d74911261052a135444cbeced49346bb17f/lib/fitbit-weather/companion.js#L58)

----------
> Should it be fetched regularly and cached or fetched only when asked to display ? (Caching the previous while waiting for fetch then update the UI)

It's up to you.
The fitbit-weather module can use caching if you specify a [maxium age](https://github.com/gregoiresage/fitbit-weather#api) parameter.
In my clock face I use a combination of an hour caching data + trying to fetch new data only when asked to display. But it can be edited in the clock face's settings.
Caching is useful to avoid calling too many times DarkSky's API as it's free only for the first 1k requests.

> [...] if the Versa calls it, I supposed the phone would fetch then send through websocket the result, but I'm not sure about it.

Yes, this is exactly the process. Same goes for the Ionic.

Do you have a Versa ?
Because I got some users which have trouble with displaying the weather's data and I think it's because the Versa has no built-in GPS (these users also had iOS devices so it may be permissions' issues).
The Versa has to use the mobile's GPS where as the Ionic can use its own built-in.
So I've allowed users to [manually configure](https://github.com/rootasjey/metrix-fitbit/blob/83765d74911261052a135444cbeced49346bb17f/settings/index.jsx#L62) their location.

Let me know if you have more questions.
I struggled with this too when I first implemented fetch data.

PS: I like & used your clock face :)

**Comment by Gby56** [comment]
Thanks you SO MUCH for all the explanation, I finally figured out fitbit-weather and just popped it in and it worked right up :)

I just finished a simple clockface, by recycling [FibitArcs](https://github.com/Gby56/FitbitArcs), that's why I forked it, I'll push my code once I finish testing out the battery life and usability.

I'm using the Versa, and yes, the location might take a little while to show up, so the first minute or so Darksky isn't working for example. I'm thinking of reusing the Statemanager from FitbitArcs and implement stocks, and other APIs for versatility.

I also added an automatic switch between the different values (steps, cal, HR, weather), to keep the UI minimal and cleaner.
